infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msgHandlers/LogReplicationServer.java

```diff
@@ -245,89 +277,82 @@ private void handleMetadataRequest(@Nonnull RequestMsg request,
-                    .setSinkClusterId(localClusterId)
-                    .setSubscriber(SessionManager.getDefaultSubscriber())
-                    .build();
-        } else {
-            session = request.getHeader().getSession();
+                    request.getHeader().getRequestId());
+            sendLeadershipLoss(request, router);
         }
-
-        return session;
-    }
-
-    private ResponseMsg getMetadataResponse(RequestMsg request, ReplicationMetadata metadata) {
-
-        LogReplicationMetadataResponseMsg metadataMsg = LogReplicationMetadataResponseMsg.newBuilder()
-                .setTopologyConfigID(metadata.getTopologyConfigId())
-                .setVersion(metadata.getVersion())
-                .setSnapshotStart(metadata.getLastSnapshotStarted())
-                .setSnapshotTransferred(metadata.getLastSnapshotTransferred())
-                .setSnapshotApplied(metadata.getLastSnapshotApplied())
-                .setLastLogEntryTimestamp(metadata.getLastLogEntryBatchProcessed())
-                .build();
-        ResponsePayloadMsg payload = ResponsePayloadMsg.newBuilder()
-                .setLrMetadataResponse(metadataMsg)
-                .build();
-
-        return getResponseMsg(getHeaderMsg(request.getHeader()), payload);
     }
 
     /**
      * Given a leadership request message, send back a
      * response indicating our current leadership status.
      *
      * @param request the leadership request message
-     * @param ctx     the context which enables a {@link ChannelHandler} to interact with its
-     *                {@link ChannelPipeline} and other handlers
      * @param router  router used for sending back the response
      */
-    @RequestHandler(type = PayloadCase.LR_LEADERSHIP_QUERY)
-    private void handleLogReplicationQueryLeadership(@Nonnull RequestMsg request,
-                                                     @Nonnull ChannelHandlerContext ctx,
-                                                     @Nonnull IServerRouter router) {
+    @LogReplicationRequestHandler(requestType = LR_LEADERSHIP_QUERY)
+    private void  handleLeadershipQuery(RequestMsg request, ResponseMsg res,
+                                                     @Nonnull IClientServerRouter router) {
         log.debug("Log Replication Query Leadership Request received by Server.");
         HeaderMsg responseHeader = getHeaderMsg(request.getHeader());
         ResponseMsg response = getLeadershipResponse(responseHeader, isLeader.get(), localNodeId);
-        router.sendResponse(response, ctx);
+        router.sendResponse(response);
+    }
+
+    /**
+     * Handle an ACK from Log Replication server.
+     *
+     * @param response The ack message
+     * @param router   A reference to the router
+     */
+    @LogReplicationResponseHandler(responseType = LR_ENTRY_ACK)
+    private void handleAck(RequestMsg req, ResponseMsg response,
```

Comment:
we can...but it will have more if-else conditions and adding maybe a couple of more methods one for request type and another for response type.

I was thinking that will be more code branching than strictly needed?
But please let me know your thoughts